[NDK][NTOS][NTDLL][KERNEL32] Fix read/write of KSYSTEM_TIME

tkreuzer opened the pull request:
## Purpose

Fix KiWriteSystemTime and move it to NDK. The previous implementation of KiWriteSystemTime was broken and updated the fields in the wrong order. Before that it was right for SystemTime and wrong for InterruptTime. ExpSetTimeZoneInformation had it wrong for the TimeZoneBias.
Add KiReadSystemTime to read KSYSTEM_TIME values correctly, instead of doing it manually (and partly wrongly) all over the place.

## Testbot runs (Filled in by Devs)

- ✅ KVM x86: https://reactos.org/testman/compare.php?ids=104214,104217,104227,104230
- ✅ KVM x64: https://reactos.org/testman/compare.php?ids=104207,104216,104226,104229

New run (2026-04-23):
- [x] KVM x86: https://reactos.org/testman/compare.php?ids=106969,106970,106974
  riched32:editor and rpcrt4:cstub did run without regressions, but the debug output was messed up.
- [x] KVM x64: https://reactos.org/testman/compare.php?ids=106962,106963,106975
  user32 tests are not regressions, but fixes from a previous merge that wasn't built, so the comparison is to an older version.

tkreuzer committed:
[NDK][NTOS][NTDLL][KRNEL32] Fix read/write of KSYSTEM_TIME

Fix KiWriteSystemTime and move it to NDK. The previous implementation of KiWriteSystemTime was broken and updated the fields in the wrong order. Before that it was right for SystemTime and wrong for InterruptTime. ExpSetTimeZoneInformation had it wrong for the TimeZoneBias.
Add KiReadSystemTime to read KSYSTEM_TIME values correctly, instead of doing it manually (and partly wrongly) all over the place.

tkreuzer committed:
[NDK][NTDLL][KERNEL32] Improve tick count to ms calculation

Use an inline function to do the calculation (optimized for 32/64 bit).

Extravert-ir commented on dll/ntdll/rtl/libsupp.c:
```diff
@@ -1221,27 +1221,19 @@ ULONG
 
 #ifdef _WIN64
-    TickCount.QuadPart = *((volatile ULONG64*)&SharedUserData->TickCount);
+    return (TickCount.QuadPart * SharedUserData->TickCountMultiplier) >> 24;
```
Can't you use the same code for both 32bits and 64bits? Does this have a different semantic depending on bitness?

tkreuzer replied:
Semantically it should be the same. I'll have a look.

Extravert-ir replied:
I believe that a small performance improvement does not worth completely destroying the ability to understand what's happening here. The compiler supports 64bit math and all that  trickery with special macros or manually manipulating parts of integer should be left in 2005.

So my suggestion would be to just leave `(TickCount.QuadPart * SharedUserData->TickCountMultiplier) >> 24` here without any `#ifdef` and do that in other similar places.
Switching `LARGE_INTEGER` to `INT64` would be also nice where possible...

tkreuzer replied:
I replaced it with an inline function.

disean commented on dll/ntdll/rtl/libsupp.c:
```diff
@@ -1221,27 +1221,19 @@ ULONG
-    }
-#endif
-
+    ULONG TickCountMultiplier = SharedUserData->TickCountMultiplier;
```
I wonder if it might be possible to use `Int64ShrlMod32` intrinsic like below
https://git.reactos.org/?p=reactos.git;a=blob;f=win32ss/gdi/eng/engmisc.c;hb=2e85425c5330661dcc4fef68904e6675cafc0a6a#l298

tkreuzer replied:
Or maybe we should just use native math on x86 and let the compiler do whatever it needs to do. I might do some tests. But I will not address that in this PR.

tkreuzer commented:
I did some investigation regarding the math for the tick count multiplication. On x86 the intrinsics give the best code:
https://godbolt.org/z/Gsr5Y7b4n

HBelusca commented on ntoskrnl/ke/time.c:
```diff
@@ -66,7 +51,7 @@ KeUpdateSystemTime(IN PKTRAP_FRAME TrapFrame,
 {
     PKPRCB Prcb = KeGetCurrentPrcb();
-    ULARGE_INTEGER CurrentTime, InterruptTime;
+    LARGE_INTEGER CurrentTime, InterruptTime;
```
Why now using signed times?

tkreuzer replied:
because they are all signed

HBelusca commented on sdk/include/ndk/kefuncs.h:
```diff
@@ -394,6 +394,58 @@ KeRaiseUserException(
+
+FORCEINLINE
+LARGE_INTEGER
+KiReadSystemTime(
```
Suggestion: use a pointer to return the read time, instead of using an intermediate local variable? (like in the KeQueryTickCount function/WDK macro)
```c
FORCEINLINE
VOID
KiReadSystemTime(
    _Out_ PLARGE_INTEGER Time,
    _In_ volatile const KSYSTEM_TIME *SystemTime)
```

tkreuzer replied:
Why?

HBelusca replied:
> Why?

or like KiWriteSystemTime below.

tkreuzer commented on sdk/include/ndk/kefuncs.h:
```diff
@@ -394,6 +394,58 @@ KeRaiseUserException(
+       if the address isn't 64 bit aligned. But as long as the entire 64 bits
+       are within a single cache line, we should be good (on x64 at least,
+       when it comes to ARM64, all bets are off) This is also what Windows does. */
+    *(LONGLONG*)SystemTime = NewTime.QuadPart;
```
LARGE_INTEGER is signed. And `LONGLONG` is exactly the same as `LONG64`. And the `volatile` is pointless.

HBelusca replied:
it's pointless only when writing to the variable?

tkreuzer commented on dll/ntdll/rtl/libsupp.c:
```diff
@@ -1221,27 +1221,19 @@ ULONG
 RtlGetTickCount(VOID)
 {
-    ULARGE_INTEGER TickCount;
+    LARGE_INTEGER TickCount;
```
Ask Microsoft. They made system time signed.

tkreuzer commented on sdk/include/ndk/kefuncs.h:
```diff
@@ -394,6 +394,58 @@ KeRaiseUserException(
+    {
+        Time.HighPart = SystemTime->High1Time;
+        Time.LowPart = SystemTime->LowPart;
+        if (Time.HighPart == SystemTime->High2Time) break;
```
alright...

tkreuzer commented on sdk/include/ndk/kefuncs.h:
```diff
@@ -394,6 +394,58 @@ KeRaiseUserException(
+
+#ifdef _WIN64
+    /* Do a single atomic read */
+    Time.QuadPart = *(volatile ULONG64*)SystemTime;
```
Technically it's all the same thing here. I'll change it to all signed, because that is how MS decided to define system time.

HBelusca replied:
> I'll change it to all signed, because that is how MS decided to define system time.

Ok thanks this also answers my other question.
But using signed vs. unsigned time won't cause problems when its value goes over 0x80000000`00000000 ?

tkreuzer committed:
[NDK][XDK] Improve SharedUserData related definitions

- Move definition of USER_SHARED_DATA from pstypes.h to ketypes.h
- Add missing definition of KI_USER_SHARED_DATA for ARM64
- Remove definition of KI_USER_SHARED_DATA_PHYSICAL

HBelusca reviewed (approved):
minor thoughts, otherwise LGTM
